# Data Mining 

The main aim of this file is to extract valuable insights from the data.  

Thereafter, we will use the data insights to create and set up a dash dashboard



In [1]:
!pip install pandas


In [2]:
#loading the data

#importing pandas
import pandas as pd

#loading the files
file1= pd.read_csv("online_retail_2009.csv")
file2= pd.read_csv("online_retail.csv")

#combining the files
retail_data = pd.concat([file1, file2], ignore_index=True)
retail_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


In [3]:
#previewing the data
retail_data.head(15)

Invoice StockCode                          Description  Quantity  \
0   489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1   489434    79323P                   PINK CHERRY LIGHTS        12   
2   489434    79323W                  WHITE CHERRY LIGHTS        12   
3   489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4   489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
5   489434     22064           PINK DOUGHNUT TRINKET POT         24   
6   489434     21871                  SAVE THE PLANET MUG        24   
7   489434     21523   FANCY FONT HOME SWEET HOME DOORMAT        10   
8   489435     22350                            CAT BOWL         12   
9   489435     22349       DOG BOWL , CHASING BALL DESIGN        12   
10  489435     22195         HEART MEASURING SPOONS LARGE        24   
11  489435     22353   LUNCHBOX WITH CUTLERY FAIRY CAKES         12   
12  489436    48173C                DOOR MAT BLACK FLOCK         10   
13  489436     21755             LOVE BUILDING BLOCK WORD        18   
14  489436     21754             HOME BUILDING BLOCK WORD         3   

       InvoiceDate  Price  Customer ID         Country  
0   12/1/2009 7:45   6.95      13085.0  United Kingdom  
1   12/1/2009 7:45   6.75      13085.0  United Kingdom  
2   12/1/2009 7:45   6.75      13085.0  United Kingdom  
3   12/1/2009 7:45   2.10      13085.0  United Kingdom  
4   12/1/2009 7:45   1.25      13085.0  United Kingdom  
5   12/1/2009 7:45   1.65      13085.0  United Kingdom  
6   12/1/2009 7:45   1.25      13085.0  United Kingdom  
7   12/1/2009 7:45   5.95      13085.0  United Kingdom  
8   12/1/2009 7:46   2.55      13085.0  United Kingdom  
9   12/1/2009 7:46   3.75      13085.0  United Kingdom  
10  12/1/2009 7:46   1.65      13085.0  United Kingdom  
11  12/1/2009 7:46   2.55      13085.0  United Kingdom  
12  12/1/2009 9:06   5.95      13078.0  United Kingdom  
13  12/1/2009 9:06   5.45      13078.0  United Kingdom  
14  12/1/2009 9:06   5.95      13078.0  United Kingdom

The StockCode has five numbers or five numbers and a letter. 

Finding out if there are other alternatives

In [4]:
#filtering data with five numbers only using regular expressions with the 'str.match() method provided by pandas

five_numbers = retail_data[retail_data['StockCode'].str.match(r'^\d{5}$')]
five_numbers.info()


<class 'pandas.core.frame.DataFrame'>
Index: 932385 entries, 0 to 1067369
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      932385 non-null  object 
 1   StockCode    932385 non-null  object 
 2   Description  929519 non-null  object 
 3   Quantity     932385 non-null  int64  
 4   InvoiceDate  932385 non-null  object 
 5   Price        932385 non-null  float64
 6   Customer ID  729813 non-null  float64
 7   Country      932385 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 64.0+ MB


In [5]:
#filtering data with numbers and letter  using regular expressions with the 'str.match() method provided by pandas
num_letters = retail_data[retail_data['StockCode'].str.match('^\d{5}[a-zA-Z]{1,2}$')]
num_letters.info()

<class 'pandas.core.frame.DataFrame'>
Index: 128892 entries, 1 to 1067317
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      128892 non-null  object 
 1   StockCode    128892 non-null  object 
 2   Description  127439 non-null  object 
 3   Quantity     128892 non-null  int64  
 4   InvoiceDate  128892 non-null  object 
 5   Price        128892 non-null  float64
 6   Customer ID  90813 non-null   float64
 7   Country      128892 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 8.9+ MB


In [6]:
num_letters.head(10)


Invoice StockCode                          Description  Quantity  \
1   489434    79323P                   PINK CHERRY LIGHTS        12   
2   489434    79323W                  WHITE CHERRY LIGHTS        12   
12  489436    48173C                DOOR MAT BLACK FLOCK         10   
23  489436    35004B          SET OF 3 BLACK FLYING DUCKS        12   
28  489436    84596F         SMALL MARSHMALLOWS PINK BOWL         8   
29  489436    84596L       BISCUITS SMALL BOWL LIGHT BLUE         8   
42  489437    84507B           STRIPES DESIGN MONKEY DOLL         6   
46  489437    84970S    HANGING HEART ZINC T-LIGHT HOLDER        12   
61  489438    84031A   CHARLIE+LOLA RED HOT WATER BOTTLE         56   
62  489438    84031B  CHARLIE LOLA BLUE HOT WATER BOTTLE         56   

       InvoiceDate  Price  Customer ID         Country  
1   12/1/2009 7:45   6.75      13085.0  United Kingdom  
2   12/1/2009 7:45   6.75      13085.0  United Kingdom  
12  12/1/2009 9:06   5.95      13078.0  United Kingdom  
23  12/1/2009 9:06   4.65      13078.0  United Kingdom  
28  12/1/2009 9:06   1.25      13078.0  United Kingdom  
29  12/1/2009 9:06   1.25      13078.0  United Kingdom  
42  12/1/2009 9:08   2.55      15362.0  United Kingdom  
46  12/1/2009 9:08   0.85      15362.0  United Kingdom  
61  12/1/2009 9:24   3.00      18102.0  United Kingdom  
62  12/1/2009 9:24   3.00      18102.0  United Kingdom

When the Stockcode has a letter(s) at the end, it a variant of a product. 

In [7]:
#finding if there are more variants  using regular expressions with the 'str.match() method provided by pandas
other_formats = retail_data[~retail_data['StockCode'].str.match('^\d{5}$|^\d{5}[a-zA-Z]{1,2}$')]
other_formats.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6094 entries, 89 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Invoice      6094 non-null   object 
 1   StockCode    6094 non-null   object 
 2   Description  6031 non-null   object 
 3   Quantity     6094 non-null   int64  
 4   InvoiceDate  6094 non-null   object 
 5   Price        6094 non-null   float64
 6   Customer ID  3738 non-null   float64
 7   Country      6094 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 428.5+ KB


In [8]:
other_formats.head(15)

Invoice StockCode        Description  Quantity      InvoiceDate   Price  \
89     489439      POST            POSTAGE         3   12/1/2009 9:28   18.00   
126    489444      POST            POSTAGE         1   12/1/2009 9:55  141.00   
173    489447      POST            POSTAGE         1  12/1/2009 10:10  130.00   
625    489526      POST            POSTAGE         6  12/1/2009 11:50   18.00   
735   C489535         D           Discount        -1  12/1/2009 12:11    9.00   
736   C489535         D           Discount        -1  12/1/2009 12:11   19.00   
927   C489538      POST            POSTAGE        -1  12/1/2009 12:18    9.58   
1244   489557      POST            POSTAGE         4  12/1/2009 12:52   18.00   
2377   489597  DCGS0058   MISO PRETTY  GUM         1  12/1/2009 14:28    0.83   
2378   489597  DCGS0068  DOGS NIGHT COLLAR         1  12/1/2009 14:28    8.65   
2379   489597       DOT     DOTCOM POSTAGE         1  12/1/2009 14:28  647.19   
2539   489600       DOT     DOTCOM POSTAGE         1  12/1/2009 14:43   55.96   
2551   489601       DOT     DOTCOM POSTAGE         1  12/1/2009 14:44   68.39   
2571   489602       DOT     DOTCOM POSTAGE         1  12/1/2009 14:45   59.35   
2619   489603       DOT     DOTCOM POSTAGE         1  12/1/2009 14:46   42.39   

      Customer ID         Country  
89        12682.0          France  
126       12636.0             USA  
173       12362.0         Belgium  
625       12533.0         Germany  
735       15299.0  United Kingdom  
736       15299.0  United Kingdom  
927       15796.0  United Kingdom  
1244      12490.0          France  
2377          NaN  United Kingdom  
2378          NaN  United Kingdom  
2379          NaN  United Kingdom  
2539          NaN  United Kingdom  
2551          NaN  United Kingdom  
2571          NaN  United Kingdom  
2619          NaN  United Kingdom

In [9]:
#evaluating the number of unique items per column

print(other_formats.nunique())

Invoice        5524
StockCode        63
Description      45
Quantity        101
InvoiceDate    5222
Price          2320
Customer ID    1134
Country          33
dtype: int64


In [10]:
#finding the missing values per column 

print(other_formats.isnull().sum())

Invoice           0
StockCode         0
Description      63
Quantity          0
InvoiceDate       0
Price             0
Customer ID    2356
Country           0
dtype: int64


In [11]:
#print out any unique StockCode and Description Combination in the other 

#droping rowss based on the two columns 
other_formats_clean = other_formats.dropna(subset=['StockCode', 'Description'])
other_formats_codes = other_formats_clean.drop_duplicates(subset=['StockCode','Description'])

print(other_formats_codes)

        Invoice     StockCode                          Description  Quantity  \
89       489439          POST                              POSTAGE         3   
735     C489535             D                             Discount        -1   
2377     489597      DCGS0058                     MISO PRETTY  GUM         1   
2378     489597      DCGS0068                    DOGS NIGHT COLLAR         1   
2379     489597           DOT                       DOTCOM POSTAGE         1   
2697     489609             M                               Manual         1   
8371     490074      DCGS0004           HAYNES CAMPER SHOULDER BAG         1   
8373     490074      DCGS0076         SUNJAR LED NIGHT NIGHT LIGHT         1   
9292     490127            C2                             CARRIAGE         1   
18410   C490943  BANK CHARGES                         Bank Charges        -1   
20249    491041      DCGS0003                  BOXED GLASS ASHTRAY         1   
27994    491725       TEST001           

In [12]:
# I want to know the number of Unique values in the other formats 

#grouping the data
stock_counts =  other_formats.groupby(['StockCode', 'Description']).size().reset_index(name='Count')

#sorting the data in descending order
sorted_stock_counts = stock_counts.sort_values(by = 'Count', ascending = False)

print(sorted_stock_counts)

       StockCode                          Description  Count
38          POST                              POSTAGE   2115
35           DOT                       DOTCOM POSTAGE   1444
36             M                               Manual   1421
9             C2                             CARRIAGE    279
11             D                             Discount    177
39             S                              SAMPLES    104
8   BANK CHARGES                         Bank Charges     96
5      AMAZONFEE                           AMAZON FEE     43
2         ADJUST  Adjustment by john on 26/01/2010 16     38
18      DCGS0058                     MISO PRETTY  GUM     30
3         ADJUST  Adjustment by john on 26/01/2010 17     26
44  gift_0001_20   Dotcomgiftshop Gift Voucher £20.00     26
46  gift_0001_30   Dotcomgiftshop Gift Voucher £30.00     24
33     DCGSSGIRL                      GIRLS PARTY BAG     23
31      DCGSSBOY                       BOYS PARTY BAG     21
37          PADS        

Removing all the rows below samples, because there is nobody to explain why the format is different. 

I will drop them because I suppose that this were old products, and considering they are few observations, they will not have such an impact. 

However, it would have been useful for the stalkholder to expound why the values are different. 



In [13]:
retail_data_filtered = retail_data[retail_data['StockCode'].str.match('^\d{5}$|^\d{5}[a-zA-Z]{1,2}$') | 
                                   retail_data['StockCode'].isin(['POST', 'DOT', 'M', 'C2','D'])]

retail_data_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1066725 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1066725 non-null  object 
 1   StockCode    1066725 non-null  object 
 2   Description  1062394 non-null  object 
 3   Quantity     1066725 non-null  int64  
 4   InvoiceDate  1066725 non-null  object 
 5   Price        1066725 non-null  float64
 6   Customer ID  824209 non-null   float64
 7   Country      1066725 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 73.2+ MB


Reduced the number of observations from by 646 from 1067371 to 1066725

In [14]:
#Evaluating the number of missing values 
retail_data_filtered.isnull().sum()


Invoice             0
StockCode           0
Description      4331
Quantity            0
InvoiceDate         0
Price               0
Customer ID    242516
Country             0
dtype: int64

We have StockCodes, but we have missing values on the Description Column. Therefore, I would like us to fill the missing values(Description column) by scanning the Stockcodes. 

In [15]:
#creating a copy of the dataframe
retail_copy = retail_data_filtered.copy()

# Grouping the data for each StockCode (creates a sub dataframe)
# The transform method applies a particular transformation to each description column in each group
#x.ffill() forward fills and bfill() backward fills the NaN values in the Description column 
retail_copy['Description'] = retail_copy.groupby('StockCode')['Description'].transform(lambda x: x.ffill().bfill())

retail_data_filtered = retail_copy

retail_data_filtered.isnull().sum()

Invoice             0
StockCode           0
Description       341
Quantity            0
InvoiceDate         0
Price               0
Customer ID    242516
Country             0
dtype: int64

This has reduced the number of missing values(Description column) from 4331 to 341




In [16]:
missing_descriptions = retail_data_filtered[retail_data_filtered['Description'].isnull()]
#missing_descriptions.head(50)
missing_descriptions.tail(20)

Invoice StockCode Description  Quantity      InvoiceDate  Price  \
194005  507870    85231K         NaN      -145  5/11/2010 14:58    0.0   
194359  507900    84706E         NaN       -40  5/11/2010 17:07    0.0   
194527  507938     16255         NaN       -76  5/12/2010 10:12    0.0   
194545  507962    85086B         NaN       -33  5/12/2010 10:36    0.0   
199647  508519     35900         NaN       -28   5/17/2010 9:44    0.0   
200898  508658    35908C         NaN       -59  5/17/2010 16:17    0.0   
206075  509283     21273         NaN        -2  5/21/2010 11:42    0.0   
206175  509320     20824         NaN       -74  5/21/2010 12:10    0.0   
210984  509776     20690         NaN        -3  5/25/2010 16:50    0.0   
211084  509840    71406B         NaN       -81  5/25/2010 17:51    0.0   
221111  510905     90117         NaN       -23   6/4/2010 10:14    0.0   
221113  510906     90107         NaN       -48   6/4/2010 10:15    0.0   
264918  515002     21048         NaN        -2   7/7/2010 15:35    0.0   
264919  515001     21920         NaN        -2   7/7/2010 15:35    0.0   
274021  516009     85074         NaN      -120   7/16/2010 9:41    0.0   
274022  516008     85073         NaN      -599   7/16/2010 9:41    0.0   
341995  522781    84247D         NaN        -3  9/16/2010 14:38    0.0   
431435  530600    85017D         NaN       -36  11/3/2010 15:17    0.0   
634150  545566    37477D         NaN        -9   3/3/2011 16:14    0.0   
872310  567207    35592T         NaN         4  9/19/2011 11:01    0.0   

        Customer ID         Country  
194005          NaN  United Kingdom  
194359          NaN  United Kingdom  
194527          NaN  United Kingdom  
194545          NaN  United Kingdom  
199647          NaN  United Kingdom  
200898          NaN  United Kingdom  
206075          NaN  United Kingdom  
206175          NaN  United Kingdom  
210984          NaN  United Kingdom  
211084          NaN  United Kingdom  
221111          NaN  United Kingdom  
221113          NaN  United Kingdom  
264918          NaN  United Kingdom  
264919          NaN  United Kingdom  
274021          NaN  United Kingdom  
274022          NaN  United Kingdom  
341995          NaN  United Kingdom  
431435          NaN  United Kingdom  
634150          NaN  United Kingdom  
872310          NaN  United Kingdom

Investigating this further, there is a trend of having zero values in the price column and negative values for the quantity column. 

Therefore, let's evaluate the number of unique values for each column

In [17]:
#evaluating unique values 
missing_descriptions.nunique()

Invoice        341
StockCode      335
Description      0
Quantity        99
InvoiceDate    280
Price            1
Customer ID      0
Country          1
dtype: int64

Having confirmed that the rows missing a Description also have 0.0 in the price. 

The proposal would have to seek explanations on the same, but since that is not feasible, the next best action is to drop this values. 

This will reduce the number of observations to 1066384

In [18]:
retail_data_filtered= retail_data_filtered.dropna(subset=['Description'])

retail_data_filtered.nunique()

Invoice        52952
StockCode       4912
Description     5659
Quantity        1053
InvoiceDate    47233
Price           2581
Customer ID     5922
Country           43
dtype: int64

Before evaluating the price point, there are more descriptions than existing StockCode. We evaluate the StockCodes with more than one description 



In [19]:
#group the data using StockCodes, and declare the level of aggregation as unique description 

grouped = retail_data_filtered.groupby('StockCode').agg({'Description':'nunique'})

#filter out StockCodes with more than one description
multiple_descriptions = grouped[grouped['Description']>1].index

#creating a subset of the data 
subset = retail_data_filtered[retail_data_filtered['StockCode'].isin(multiple_descriptions)]

#dropping duplicates based on StockCode and Description 

subset = subset.drop_duplicates(subset=['StockCode', 'Description'])

subset = subset.sort_values(by='StockCode')

subset.head(15)

Invoice StockCode                      Description  Quantity  \
6339     489879     10080         GROOVY CACTUS INFLATABLE         1   
979826   575518     10080                            check        22   
2944     489640     10120                     DOGGY RUBBER        60   
194372   507913     10120              Zebra invcing error     -9000   
475      489522     10133     COLOURING PENCILS BROWN TUBE         6   
1000108  576843     10133                          damaged       -82   
100429   498888    15056N              wedding co returns?      -700   
614      489526    15056N        EDWARDIAN PARASOL NATURAL         6   
67142    495362    15058A  BLUE WHITE SPOTS GARDEN PARASOL         2   
717751   553394    15058A                        wet/rusty       -30   
548898   538200    15058A     BLUE POLKADOT GARDEN PARASOL         1   
54894    494382    15058B  PINK WHITE SPOTS GARDEN PARASOL         1   
461044   532993    15058B     PINK POLKADOT GARDEN PARASOL         2   
67187    495362    15058C  ICE CREAM DESIGN GARDEN PARASOL         2   
717750   553395    15058C                        wet/rusty       -30   

              InvoiceDate  Price  Customer ID         Country  
6339      12/2/2009 16:02   0.85      17660.0  United Kingdom  
979826   11/10/2011 10:53   0.00          NaN  United Kingdom  
2944      12/1/2009 16:17   0.21      17345.0  United Kingdom  
194372    5/11/2010 17:16   0.00          NaN  United Kingdom  
475       12/1/2009 11:45   0.85      15998.0  United Kingdom  
1000108  11/16/2011 15:32   0.00          NaN  United Kingdom  
100429    2/23/2010 13:08   0.00          NaN  United Kingdom  
614       12/1/2009 11:50   5.95      12533.0         Germany  
67142     1/24/2010 12:12   7.95      15727.0  United Kingdom  
717751    5/16/2011 16:47   0.00          NaN  United Kingdom  
548898   12/10/2010 11:11   7.95      17912.0  United Kingdom  
54894     1/13/2010 17:51  25.49          NaN  United Kingdom  
461044   11/15/2010 14:26   7.95      14156.0            EIRE  
67187     1/24/2010 12:12   7.95      15727.0  United Kingdom  
717750    5/16/2011 16:47   0.00          NaN  United Kingdom

A very intresting development, some of the StockCodes have repeated descriptions showing negative changes to the stock, be it returns, damages, etc. While there are spelling errors or different products(Description) that share the same StockCode

Separate the dataset into two, one with the negative QTY values and one with positive values. 



In [20]:
#filtering the data with positive qty
positive_qty = retail_data_filtered[retail_data_filtered['Quantity']>0]

#filtering the data with negative qty
negative_qty = retail_data_filtered[retail_data_filtered['Quantity']<=0]

#explore positive qty dataframe more
positive_qty.info()



<class 'pandas.core.frame.DataFrame'>
Index: 1044010 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1044010 non-null  object 
 1   StockCode    1044010 non-null  object 
 2   Description  1044010 non-null  object 
 3   Quantity     1044010 non-null  int64  
 4   InvoiceDate  1044010 non-null  object 
 5   Price        1044010 non-null  float64
 6   Customer ID  805521 non-null   float64
 7   Country      1044010 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 71.7+ MB


In [21]:
#finding unique items per column 
positive_qty.nunique()

Invoice        41764
StockCode       4892
Description     5452
Quantity         564
InvoiceDate    38337
Price           2221
Customer ID     5870
Country           43
dtype: int64

The data still needs more cleaning 

Before we continue with the cleaning, we might as well explore the negative_qty dataframe 

In [22]:
negative_qty.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22374 entries, 178 to 1067178
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Invoice      22374 non-null  object 
 1   StockCode    22374 non-null  object 
 2   Description  22374 non-null  object 
 3   Quantity     22374 non-null  int64  
 4   InvoiceDate  22374 non-null  object 
 5   Price        22374 non-null  float64
 6   Customer ID  18688 non-null  float64
 7   Country      22374 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 1.5+ MB


In [23]:
negative_qty.nunique()

Invoice        11188
StockCode       3769
Description     4069
Quantity         489
InvoiceDate    10329
Price            809
Customer ID     2552
Country           36
dtype: int64

In [24]:
#close evaluation of this values 
negative_qty_subset = negative_qty.drop_duplicates(subset=['StockCode', 'Description'])

negative_qty_subset = negative_qty_subset.sort_values(by='StockCode')

negative_qty_subset.head(15)

Invoice StockCode                   Description  Quantity  \
148285   C503527     10002   INFLATABLE POLITICAL GLOBE       -200   
81500     496879     10109          BENDY COLOUR PENCILS        -4   
194372    507913     10120           Zebra invcing error     -9000   
118533   C500705     10120                  DOGGY RUBBER       -30   
83382    C497151    10123C         HEARTS WRAPPING TAPE         -6   
359656   C524235    10123G       ARMY CAMO WRAPPING TAPE     -1104   
1000108   576843     10133                       damaged       -82   
170277   C505529     10133  COLOURING PENCILS BROWN TUBE       -10   
170278   C505529     10134  COLOURING PENCILS BROWN TUBE       -10   
170276   C505529     10135  COLOURING PENCILS BROWN TUBE       -10   
17672    C490807     10138     ASSORTED COLOUR JUMBO PEN       -96   
18398    C490942     11001   ASSTD DESIGN RACING CAR PEN       -16   
436898    531140     15002                       rex use     -1000   
98512    C498604     15034   PAPER POCKET TRAVELING FAN        -15   
191662   C507599     15036     ASSORTED COLOURS SILK FAN       -24   

              InvoiceDate  Price  Customer ID         Country  
148285     4/1/2010 13:13   0.72      16019.0  United Kingdom  
81500      2/4/2010 13:47   0.00          NaN  United Kingdom  
194372    5/11/2010 17:16   0.00          NaN  United Kingdom  
118533     3/9/2010 13:46   0.21      14748.0  United Kingdom  
83382      2/5/2010 15:59   0.65      17412.0  United Kingdom  
359656    9/28/2010 11:02   0.06      14277.0          France  
1000108  11/16/2011 15:32   0.00          NaN  United Kingdom  
170277    4/22/2010 17:37   0.85      17679.0  United Kingdom  
170278    4/22/2010 17:37   1.25      17679.0  United Kingdom  
170276    4/22/2010 17:37   1.25      17679.0  United Kingdom  
17672     12/8/2009 12:25   0.42      13091.0  United Kingdom  
18398     12/8/2009 14:04   1.69      14337.0  United Kingdom  
436898    11/5/2010 14:11   0.00          NaN  United Kingdom  
98512     2/21/2010 16:16   0.07      13081.0  United Kingdom  
191662    5/10/2010 13:26   0.75      13136.0  United Kingdom

A very intresting development, there are elements negative values that are associated with transactions such as discount, returns and they are denoted by letter "C" in the invoice. 

In [25]:
sales= retail_data_filtered[(retail_data_filtered['Quantity']>0)| (retail_data_filtered['Invoice'].str.startswith('C'))]

negative_transactions = retail_data_filtered[(retail_data_filtered['Quantity']<= 0) & (~retail_data_filtered['Invoice'].str.startswith('C'))]

sales.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1063243 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1063243 non-null  object 
 1   StockCode    1063243 non-null  object 
 2   Description  1063243 non-null  object 
 3   Quantity     1063243 non-null  int64  
 4   InvoiceDate  1063243 non-null  object 
 5   Price        1063243 non-null  float64
 6   Customer ID  824209 non-null   float64
 7   Country      1063243 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 73.0+ MB


In [26]:
negative_transactions.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3141 entries, 263 to 1064386
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Invoice      3141 non-null   object 
 1   StockCode    3141 non-null   object 
 2   Description  3141 non-null   object 
 3   Quantity     3141 non-null   int64  
 4   InvoiceDate  3141 non-null   object 
 5   Price        3141 non-null   float64
 6   Customer ID  0 non-null      float64
 7   Country      3141 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 220.9+ KB


In [27]:
negative_transactions.nunique()

Invoice        3077
StockCode      2161
Description    1946
Quantity        434
InvoiceDate    2430
Price             1
Customer ID       0
Country           1
dtype: int64

In [28]:
negative_transactions.tail(15)

Invoice StockCode                      Description  Quantity  \
1060595  581190    84801A                            check      -102   
1060782  581198     22025                            check       -30   
1060783  581199     84581  DOG TOY WITH PINK CROCHET SKIRT        -2   
1060784  581200    84508C                            check       -21   
1060785  581201     22217                         damages?      -155   
1060788  581204     85104                  ????damages????      -355   
1060789  581205     20893                          damages       -55   
1060790  581206     21693                         mixed up       -87   
1060791  581207     21688                         mixed up      -337   
1060792  581208    72801C                            check       -10   
1060794  581210     23395                            check       -26   
1060796  581212     22578                             lost     -1050   
1060797  581213     22576                            check       -30   
1062371  581226     23090                          missing      -338   
1064386  581422     23169                          smashed      -235   

             InvoiceDate  Price  Customer ID         Country  
1060595  12/7/2011 16:55    0.0          NaN  United Kingdom  
1060782  12/7/2011 18:26    0.0          NaN  United Kingdom  
1060783  12/7/2011 18:26    0.0          NaN  United Kingdom  
1060784  12/7/2011 18:27    0.0          NaN  United Kingdom  
1060785  12/7/2011 18:30    0.0          NaN  United Kingdom  
1060788  12/7/2011 18:32    0.0          NaN  United Kingdom  
1060789  12/7/2011 18:32    0.0          NaN  United Kingdom  
1060790  12/7/2011 18:34    0.0          NaN  United Kingdom  
1060791  12/7/2011 18:34    0.0          NaN  United Kingdom  
1060792  12/7/2011 18:35    0.0          NaN  United Kingdom  
1060794  12/7/2011 18:36    0.0          NaN  United Kingdom  
1060796  12/7/2011 18:38    0.0          NaN  United Kingdom  
1060797  12/7/2011 18:38    0.0          NaN  United Kingdom  
1062371   12/8/2011 9:56    0.0          NaN  United Kingdom  
1064386  12/8/2011 15:24    0.0          NaN  United Kingdom

The negative transactions dataframe shows reduction in inventory that does not necessarily result from customer transactions. This will be analyzed to show inventory losses. 

In [29]:
#saving the negative_transactions as a csv

negative_transactions.to_csv('inventory_losses,csv', index = False)

In [30]:
#exploring the sales transactions
sales.nunique()

Invoice        49875
StockCode       4903
Description     5463
Quantity         770
InvoiceDate    45304
Price           2581
Customer ID     5922
Country           43
dtype: int64

Still, there are StockCodes with more than one description

In [31]:
#getting a list of StockCodes with more than one description 
#group the data using StockCodes, and declare the level of aggregation as unique description 

sales_grouped = sales.groupby('StockCode').agg({'Description':'nunique'})

#filter out StockCodes with more than one description
multiple_descriptions = sales_grouped[sales_grouped['Description']>1].index

#creating a subset of the data 
sales_subset = sales[sales['StockCode'].isin(multiple_descriptions)]

#dropping duplicates based on StockCode and Description 

sales_subset = sales_subset.drop_duplicates(subset=['StockCode', 'Description'])

sales_subset = sales_subset.sort_values(by='StockCode')

sales_subset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1727 entries, 6339 to 2376
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Invoice      1727 non-null   object 
 1   StockCode    1727 non-null   object 
 2   Description  1727 non-null   object 
 3   Quantity     1727 non-null   int64  
 4   InvoiceDate  1727 non-null   object 
 5   Price        1727 non-null   float64
 6   Customer ID  1202 non-null   float64
 7   Country      1727 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 121.4+ KB


In [32]:
sales_subset.head(50)

Invoice StockCode                          Description  Quantity  \
6339     489879     10080             GROOVY CACTUS INFLATABLE         1   
979826   575518     10080                                check        22   
67142    495362    15058A      BLUE WHITE SPOTS GARDEN PARASOL         2   
548898   538200    15058A         BLUE POLKADOT GARDEN PARASOL         1   
461044   532993    15058B         PINK POLKADOT GARDEN PARASOL         2   
54894    494382    15058B      PINK WHITE SPOTS GARDEN PARASOL         1   
119710   500843     16011                      ANIMAL STICKERS        31   
10462    490163     16011                      ANIMAL STICKERS        24   
4157     489800     16012           FOOD/DRINK SPUNGE STICKERS        61   
314800   520027     16012           FOOD/DRINK SPONGE STICKERS        24   
710756   552803    16151A      FLOWERS HANDBAG blue and orange         2   
365004   524749    16151A                                found        50   
181581   506609    16151A  FLOWER DES BLUE HANDBAG/ORANG HANDL         1   
49282    493985    16156L                       WRAP, CAROUSEL        25   
663541   548189    16156L                        WRAP CAROUSEL        25   
2497     489599    16161U                WRAP,SUKI AND FRIENDS        25   
258896   514367    16161U                WRAP SUKI AND FRIENDS        25   
790759   560218    16162M  alan hodge cant mamage this section        30   
202650   508903    16162M         THE KING GIFT BAG 25x24x12cm       144   
919368   570860    16168M                                found        44   
21672    491110    16168M         FUNKY MONKEY GIFT BAG MEDIUM         3   
959154   573984    16207A                                check         2   
364      489517    16207A             PINK STRAWBERRY HANDBAG          2   
109455   499897    17012F  ORIGAMI SANDLEWOOD INCENSE/CAND SET        24   
993760   576414    17012F                                check        14   
692106   550950    17091A                                found        84   
13959    490506    17091A              LAVENDER INCENSE IN TIN         1   
10248    490151    17107D  FLOWER FAIRY,5 SUMMER B'DRAW LINERS         6   
873225   567338    17107D    FLOWER FAIRY 5 SUMMER DRAW LINERS         6   
841204   564734    17107D         FLOWER FAIRY 5 DRAWER LINERS         4   
4280     489805    17109B        FLOWER FAIRY INCENSE LAVENDER         2   
113350   500248    17109B                                found        72   
10237    490151    17109D         FLOWER FAIRY INCENSE BOUQUET         6   
692101   550948    17109D                           Adjustment        14   
182998   506748    17129C                           wrong invc        70   
21857    491119    17129C    SKY BLUE COLOUR GLASS GEMS IN BAG         2   
1879     489597    17129D       CLEAR COLOUR GLASS GEMS IN BAG         1   
47737    493837    17129D             missing (wrongly coded?)       120   
272123   515809    17129F               BLUE GLASS GEMS IN BAG         8   
9714     490149    17129F          DEEP BLUE GLASS GEMS IN BAG         2   
6560     489900    17136A       BLUE STONES ON WIRE FOR CANDLE        24   
1014657  577954    17136A                                check         3   
5619     489857    18094C   WHITE/BLUE RECT CERAMIC OIL BURNER         1   
95859    498435    18094C    WHITE AND BLUE CERAMIC OIL BURNER         3   
366785   524958     20615         BLUE POLKADOT PASSPORT COVER         1   
2509     489600     20615           BLUE SPOTTY PASSPORT COVER         1   
635      489529     20622                   VIPPASSPORT COVER          1   
769197   558475     20622                  VIP PASSPORT COVER          2   
353803   523773     20652           BLUE POLKADOT LUGGAGE TAG          1   
1882     489597     20652             BLUE SPOTTY LUGGAGE TAG          1   

              InvoiceDate  Price  Customer ID         Country  
6339      12/2/2009 16:02   0.85      17660.0  United Kingdom  
979

There is another intresting development, some duplications in the StockCode that show a positive change in the inventory. The key differentiator is that they have NaN for customer and 0.0 for price. 

In addition, there are duplications in StockCodes where there are spelling errors in the Description column  